network binding plugin: DownwardAPI for device info (#281)
kubevirt/community: design-proposals/network-binding-plugin/network-binding-plugin.md

- DownwardAPI

##### Device Info

In some cases the multus status annotation
contains [device-info](https://github.com/k8snetworkplumbingwg/device-info-spec/blob/main/SPEC.md).
This info may be valuable for the binding plugin sidecar. For example
`SRIOV` and `VDPA` plugins need this info.

As the multus status annotation is not there yet on `virt-launcher` pod
creation, the suggested solution is to use downwardAPI volume and mount
it to the sidecar.
The downwardAPI volume will track a new `virt-launcher` pod

> annotation `kubevirt.io/network-device-info-map` that will contain a

Reviewer: Please consider removing map prefix, IMO it doesn't add value.
Also if tomorrow we decide to change the annotation value format (to not be map) the name would no longer be valid.

Author: The fact it is a map is part of the API. I cannot change it to be something else without breaking backward compatibility.

Reviewer 2: I do not think the meaning of `map` is universally understood.
In the `Go` programming language a key-value structure is named `map`, in other languages it is called a `dict` (Python).
In plain English it describes an are of land/sea.

Even in programming we try to avoid expressing the type of a variable in the name. We usually prefer the name to represent what the data contains and not how it contains it.

I also think the `map` is redundant.

> annotation `kubevirt.io/network-device-info-map` that will contain a

Reviewer 2: I would prefer a generalization to avoid the need to create multiple values that track different annotations. We already have a good example with the existing SR-IOV one.

It should be harmes to generalize it as long we have a proper structure and a general name.
The only disadvantage is the embedded json, but we have it now anyway.

Here is a potential structure:
Annotation name: `kubevirt.io/virt-launcher-data`
Annotation value format:
```json
{
  "network": {
    "interfaces": [
      "name": "net1"
      "networkLogicalName": <logical name>
      "device-info": {...}
    ]
  }
}
```

In any case, please explicitly define the format here.

Author: @EdDev 
How about?
```
type VirtLauncherData struct {
	networkDeviceInfoMap map[string]*networkv1.DeviceInfo
}
```

Reviewer 2: ```go
type VirtLauncherDownwardData struct {
	Network *Network `json:"network,omitempty"`
}
type Network struct {
	Interfaces []Interface `json:"interfaces,omitempty"`
}

type Interface struct {
	Name               string `json:"name"`
	NetworkLogicalName string `json:"networkLogicalName"`
	DeviceInfo         string `json:"deviceInfo,omitempty"`
}
```

Author: I agree, I prefer to keep it narrow as well.
If in the future there will be a need to extend it, it should be easy, the volume will have to point to a new annotation.
The only downside is that we will have to keep the old annotation for a while to keep backward compatibility for old launchers that are already running.

Reviewer 2: > Here we try to enable a way to pass data from virt-controller (who has access to virt-launcher pod annotation) to virt-launcher app. The suggestion to generalize the annotation may sounds appealing and future proof, it may seem like a mere annotation, but it actually become an internal API; communication channel between virt-controller and virt-launcher.

Indeed, it is a structured, stable and open for extension protocol.
It is not a path focused on one single problem that needs to be solved, it looks at the general need and defines it well.

> Currently we have similar channel, between virt-handler and launcher, using gRPC and protobuf (handler-launche-com) Unlike the `handler-launcher-com`, this annotation can easily be abused by the next dev and becoming a "go to workaround" when introducing an API change require more time and work in a tight time-frame. I am talking about cases when there is a need to pass data from virt-controller to virt-launcher and its possible using `handler-launcher-com ` / API change.

I do not accept this as an argument. Abusing a channel and protocol is always possible, it cannot be the reasoning for not establishing one.

With the same reasoning I can claim the usage of this DownwardAPI mechanism is an abuse, giving ideas to others to use the same for other means.

> Keep in mind that we struggle to keep track and apply proper review process to formal and stable APIs (hopefully it will get better with sig-api group). I am afraid this annotation will become hard to maintain and become another backward-compatibility thing to worry about.

This is true for any API change, no matter where it is placed.

> If in the future there will be a need to extend it, it should be easy, the volume will have to point to a new annotation.
> The only downside is that we will have to keep the old annotation for a while to keep backward compatibility for old launchers that are already running.

I do not think the notion of adding an API, even if internal, and admitting it may be replaced if an extension will be needed is a good direction. This implies revolution is better than evolution in a supposed-to-be stable project API.

I think the data communication channel should be formatted correctly to allow it to extend.
I disagree with narrowing its scope as argued above.
But as with the previous incident (SR-IOV DownwardAPI), I will not block the whole solution on this.

mapping between the logical network name to the `device-info` of this
network.
Very similar to the
existing [SRIOV downwardAPI](https://github.com/kubevirt/kubevirt/pull/8226)
solution which is mapping the networkName to the `pci-address` of the
device and mounting the downwardAPI volume to the `virt-launcher` pod.

> > **Note**: Old multus versions may not report the device-info.

Reviewer 2: Please specify the version and what will happen if it is missing.

Reviewer: I recall it was Multus that stated to reflect PCI info since version v3.7
https://github.com/kubevirt/community/blob/main/design-proposals/sriov-interfaces-mapping.md#example-1 
See the note in this section

Author: Thanks Or, updated the doc.

Alternatives:

- Instead of configuring in the kubevirt CR whether deviceInfo should be

>   mounted to the sidecar or not, we can always mount it in case it exists

Reviewer: > we can always mount it in case it exists in the status

This is not clear could please rephrase it to be more explicit?

  in the status. The downside in this approach is that a DownwardAPI
  volume will be added to the pod even when it is not actually needed.
- Passing `PCIDEVICE_<RESOURCE_NAME>_NET_INFO` env var from the
  `virt-launcher` to the sidecar by adding it as an annotation to the
  vmi object passed to the onDefineDomain method. The problem with this
  approach is that it will reintroduce the

>   same [bug](https://github.com/kubevirt/kubevirt/issues/7444) SRIOV

Reviewer: ```suggestion
same [bug](https://github.com/kubevirt/kubevirt/issues/7444)
```

- DownwardAPI

##### Device Info

In some cases the multus status annotation
contains [device-info](https://github.com/k8snetworkplumbingwg/device-info-spec/blob/main/SPEC.md).
This info may be valuable for the binding plugin sidecar. For example
`SRIOV` and `VDPA` plugins need this info.

As the multus status annotation is not there yet on `virt-launcher` pod
creation, the suggested solution is to use downwardAPI volume and mount
it to the sidecar.
The downwardAPI volume will track a new `virt-launcher` pod

> annotation `kubevirt.io/network-info` that will contain the `network-info`

Reviewer: Changing it to network-info immediately raise question "why passing network info and not any other info?"
I would keep the scope narrow as before and have the annotation provide device-info only.

It seems that other network info already passed to the virt-launcher app through other channels (VMI status/gRPC, guest-agent), the only missing one was the device-info that up until the SR-IOV mapping bug would be given by the device-plugin.

Author: The reason I changed it is that in the CR we define - `DownwardAPI: device-info`. Meaning the API is extendable and we may add other DownwardAPI type. I want the CR API to be align to the pod one.

of the networks that have `downwardAPI`.
Very similar to the
existing [SRIOV downwardAPI](https://github.com/kubevirt/kubevirt/pull/8226)
solution which is mapping the networkName to the `pci-address` of the
device and mounting the downwardAPI volume to the `virt-launcher` pod.

**Note**: Multus provides device-info information since version v3.7.
If an older version is used, the downward API is not supported,
the annotation will have empty values.

> 

Reviewer: I think it will be nice defining the data model of the introduced annotation value, including what info it provides (PCI address..) in which format (map, json..)

Also when this annotation and downward API volume are be added (for every VM or only for those who require plugin sidecar)

Author: > I think it will be nice defining the data model of the introduced annotation value, including what info it provides (PCI address..) in which format (map, json..)

The vDPA example has it.

> Also when this annotation and downward API volume are be added (for every VM or only for those who require plugin sidecar)

It is written - `The downwardAPI volume will track a new `virt-launcher` pod
annotation `kubevirt.io/network-info` that will contain the `network-info`
of the networks that have `downwardAPI`.`
I find the exact details of when the volume is added are implementation details and shouldn't be part of a high level design.

Reviewer: >The vDPA example has it.

Right but it is just an example I think it should be well defined.
Anyhow I am not insisting on it the proposal looks good.

> I find the exact details of when the volume is added are implementation details and shouldn't be part of a high level design.

At first I thought it could become backward-comparability concern, but it shouldn't be an issue:
On Kubevirt upgrade, for existing pods, virt-controller will just add the new annotation / new annotation value.
The annotation value wont propagate to the launcher app until the VM is migrated, 
which is fine since it cause no disruption.
New pods will be created with new annotation  volume.

Author: It is an API, so yes, a change should be considered as an API change.

  mounted to the sidecar or not, we can pass via the downward API the device
  info of all the interfaces that have it.
  The downside in this approach is that a DownwardAPI volume will be added to
  the pod even when it is not actually needed.
- Passing `PCIDEVICE_<RESOURCE_NAME>_NET_INFO` env var from the
  `virt-launcher` to the sidecar by adding it as an annotation to the
  vmi object passed to the onDefineDomain method. The problem with this
  approach is that it will reintroduce the
  same [bug](https://github.com/kubevirt/kubevirt/issues/7444)
  that the SRIOV downward API solved.
- Sharing
[CNIDeviceInfo](https://github.com/k8snetworkplumbingwg/device-info-spec/blob/main/SPEC.md#5-new-cni-plugin-arguments)
file with the sidecar. It is a very similar approach to the
downwardAPI option, the disadvantage here is that the user would have
to specify the file path in the `NetworkAttachmentDefinition`.
- Instead of using an annotation and DownwardAPI, extend the VMI interface status
API to contain the interface device info.

  Pros:
  - DownwardAPI volume is not required, the VMI is already passed to the virt-launcher
  compute and sidecar containers.

  Cons:

>   - The VMI CRD should be changed.

Reviewer 3: I do-not agree with the first two points here, see here: https://github.com/kubevirt/community/pull/281#issuecomment-2049651323

    The device infromation is exposed to users as a contract,
    while the pod annotation with the DownwardAPI is an implementation detail with no such contract.
  - It should be easy to move from the pod annotation to VMI Status in the future,
  but we cannot go the opposite way.
  - Keeping pod specific information in the VMI status may be raceful over migration.

>   - Users that have read permissions on the VMI will be able to see the device info.

Author: User that has read permission on the VMI not necessary has read permissions on the pod, therefore, with the downward API option, this kind of user won't be able to see the device info.

>   - Users that have read permissions on the VMI will be able to see the device info.

Reviewer 3: Similarly, could it be possible that some users dont have access to VMI read but have access to read pods? 

All I am saying is that depending on the cluster configuration, some could have read access and other could not. We cannot make an assumption that one solution is more secure than other because it is largely dependent on cluster configuration

>   - Users that have read permissions on the VMI will be able to see the device info.

Author: We can say that the suggested option doesn't give any user more sensitive information than it already have.
The device info was already exposed to the pod via an environment variable by the device plugin. The downward API is needed just to create the mapping between the device info to the network using it.
But the virt-launcher is already familiar with the networks and with the device info.

>   - Users that have read permissions on the VMI will be able to see the device info.

Reviewer 3: That sounds like an accurate statement, +1

    <rom enabled='no'/>
</interface>
```

## Appendix F: Concrete example for vDPA binding

Present a concrete example of the actions needed at each integration
point, considering the option suggested in the

> [design section](#design-details).

Reviewer: leftover?

Reviewer 2: What leftover do you see here?

Reviewer: lines 990-992, it doesn't say anything, just describing what it should.

### Kubevirt CR Specification

The vDPA binding is registered to the Kubevirt CR.
The plugin is then referenced by name from the VM Specification.

> **Note**: The vDPA binding does not need to perform changes in the
> pod network, therefore, there is no need for a
> network-attachment-definition.
```yaml
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  name: kubevirt
  namespace: kubevirt
spec:
  configuration:
    network:
      binding:
        vdpa:
          sidecarImage: quay.io/kubevirt/network-vdpa-binding
          downwardAPI: device-info
```

### VM Specification

```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: vmi-1-bridge
spec:
  template:
    spec:
      domain:
        devices:
          interfaces:
            - name: blue
              mac: 12:34:56:78:9a:bc
              binding:
                name: vdpa
      networks:
        - name: blue
          multus:
            networkName: vdpa-network
```

### NetworkAttachmentDefinition Specification

Not required for vDPA binding

> > **Note**: Just to clarify, a NAD is required for the vDPA CNI and
> > device plugin, but it is not in the scope of this doc. An extra NAD
> > for
> > the network binding plugin is not required.

Reviewer: ```suggestion:
Defining a net-attach-def to connect the pod to the node vDPA interface is required, the that is spesifed in the VM spec.
No additional net-attach-def required, because there is no need to configure the pod network stack for vDPA.
```

### Webhook Admitter

- The core KubeVirt admitter assures that only one method of binding

>   definition is possible: Either the old style method or the new one.

Reviewer: ```suggestion
  definition is possible: Either through the Interface API or Network Binding Plugin API.
```

Author: done.

- No addmiter is provided as part of the vDPA binding.

### Pod Definition

The VMI controller updates the pod manifest per the VMI specification.

- From the VMI spec, the following snippet is detected:

```yaml
name: blue
binding:

>   name: vdpa

Reviewer: It would be easier to understand if the at least the interface stanza is mentioned:
VMI spec.domain.devices.interfaces:
```
interfaces:
- name: blue
  binding:
    name: vdpa
```

Reviewer 2: This the format used on all other examples, I do not think it is that important to break the convention.
A different PR can suggest changing the convention to all if you would like to invest in this.

```

- As a result, the pod is updated with the following annotations & spec:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: virt-launcher-123
  annotations:
    kubevirt.io/network-device-info-map: '{
      blue:
        "device-info": {
          "type": "vdpa",
          "version": "1.1.0",
          "vdpa": {
            "parent-device": "vdpa:0000:65:00.3",
            "driver": "vhost",
            "path": "/dev/vhost-vdpa-1",
            "pci-address": "0000:65:00.3"
          }
        }
    }'
  spec:
    containers:
      - name: compute

> 

Reviewer: nit:
```suggestion
...
```

Reviewer 2: It will make the MD interpretation break and look weird, I do not recommend adding it.

Reviewer: Note that its inside "```yaml```" block, it shouldnt break.

      - name: hook-sidecar-1
        volumeMounts:

>           - name: network-device-info-map

Reviewer: Please consider removing the `map` prefix, it doesnt add value in the context of volumeMount name.

>             mountPath: /etc/pod-info

Reviewer: How about mounting the device info at its own directory? maybe `/etc/net-device-info`?

Note that the current SR-IOV mapping annotation is mounted to `/etc/pod-info`

Author: It is intentionally under the same dir. The path is different - /etc/pod-info/device-info

Reviewer: Also commented on the impl. PR 
I would store the new device-info annotation in different dir then the sriov-network-pci one.
IMO it will be easier to debug that way, also avoid potential overwrites, if tomorrow someone changes the downward API volumes render code.

> **Note**: The vDPA binding does not need to perform changes in the
> pod network, therefore, there is no need for a
> network-attachment-definition.
```yaml
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  name: kubevirt
  namespace: kubevirt
spec:
  configuration:
    network:
      binding:
        vdpa:
          sidecarImage: quay.io/kubevirt/network-vdpa-binding
          downwardAPI: device-info
```

### VM Specification

```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: vmi-1-bridge
spec:
  template:
    spec:
      domain:
        devices:
          interfaces:
            - name: blue
              mac: 12:34:56:78:9a:bc
              binding:
                name: vdpa
      networks:
        - name: blue
          multus:
            networkName: vdpa-network
```

### NetworkAttachmentDefinition Specification

Not required for vDPA binding

> **Note**: Defining a NetworkAttachmentDefinition to connect the pod to the node vDPA interface is required,
> this NetworkAttachmentDefinition is specified in the VM spec.
> No additional NetworkAttachmentDefinition is required, because there is no need to configure the pod network
> stack for vDPA.

> ### Webhook Admitter

Reviewer 2: nit: Newline is need above.

Author: done.

    <rom enabled='no'/>
</interface>
```

## Appendix F: Concrete example for vDPA binding

Present a concrete example of the actions needed at each integration
point, considering the option suggested in the
[design section](#design-details).

### Kubevirt CR Specification

The vDPA binding is registered to the Kubevirt CR.
The plugin is then referenced by name from the VM Specification.

> **Note**: The vDPA binding does not need to perform changes in the
> pod network, therefore, there is no need for a
> network-attachment-definition.
```yaml
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  name: kubevirt
  namespace: kubevirt
spec:
  configuration:
    network:
      binding:
        vdpa:
          sidecarImage: quay.io/kubevirt/network-vdpa-binding

>           downwardAPI: device-info

Reviewer 3: implementation wise, should this be a list of strings? do you anticipate that there could be more than one metadata information that could be passed to sidecar network plugins?

Author: I think that in the future there might be other data that a plugin would require but I don't think that a combination of more than one type will be required. And even if it would be (unlikely), we can always introduce a new type that represents a combination.

Reviewer 3: As an example in the future consider that the network sidecar needs  `kubevirt.io/foo` annotation along with `kubevirt.io/network-info` as metadata in sidecar. How would this API evolve?

Author: `downwardAPI: device-info-foo` or comma separated `downwardAPI: deviceinfo,foo`. Or we can introduce `downwardAPIs`. Although I don't think this kind of combination will ever be required. I cannot think of a plugin that will need information from several pod fields.  I don't even believe that anything besides deviceinfo will be required, but kept this option open.

Reviewer: This reminds me of K8S Pod status API where the IPS  filed was added once IP was not sufficient.
IMO it is not nice to add the another filed just to support multiple options, I would go with sting slice or support coma separated filed in the future (not optimal).

            - name: blue
              mac: 12:34:56:78:9a:bc
              binding:
                name: vdpa
      networks:
        - name: blue
          multus:
            networkName: vdpa-network
```

### NetworkAttachmentDefinition Specification

Not required for vDPA binding

> **Note**: Defining a NetworkAttachmentDefinition to connect the pod to the node vDPA interface is required,
> this NetworkAttachmentDefinition is specified in the VM spec.
> No additional NetworkAttachmentDefinition is required, because there is no need to configure the pod network
> stack for vDPA.

### Webhook Admitter

- The core KubeVirt admitter assures that only one method of binding
  definition is possible: Either through the Interface API or Network Binding Plugin API.
- No addmiter is provided as part of the vDPA binding.

### Pod Definition

The VMI controller updates the pod manifest per the VMI specification.

- From the VMI spec, the following snippet is detected:

```yaml
name: blue
binding:
  name: vdpa
```

- As a result, the pod is updated with the following annotations & spec:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: virt-launcher-123
  annotations:

>     kubevirt.io/network-info: '{

Reviewer 3: If the same information is available as a different annotation `k8s.v1.cni.cncf.io/network-status`, out of curiosity what problem does having a kubevirt specific annotation solve?

https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/docs/how-to-use.md#launch-pod-with-text-annotation

Author: The network-status annotation contains information about all the pod interfaces, including IPs and other confidential data.The new annotation collects from the status only the device info of the interfaces that use a plug-in that supports it, the other interface and other info besides the device-infi is not passed.
We find this approach more secure and avoiding sidecar users to get the complete pod networking information.

Author: Done, added it to the alternatives sections.

      "interfaces": [{
        "network": "blue"
        "device-info": {
          "type": "vdpa",
          "version": "1.1.0",
          "vdpa": {
            "parent-device": "vdpa:0000:65:00.3",
            "driver": "vhost",
            "path": "/dev/vhost-vdpa-1",
            "pci-address": "0000:65:00.3"
          }
        }
      }]
    }'
  spec:
    containers:
      - name: compute

      - name: hook-sidecar-1
        volumeMounts:
          - name: network-info
            mountPath: /etc/pod-info
    volumes:
      - name: network-info
        downwardAPI:
          items:

>             - path: device-info

Author: Typo, the path should be `network-info`.
